File: rnaseq_pipeline.R
#!/usr/bin/env Rscript
#
# RNA-seq Analysis Pipeline in R
# Minimal pipeline: FastQC -> STAR alignment -> featureCounts
#
# This script orchestrates bioinformatics tools using system() calls
# Similar to how many R-based analysis pipelines work in practice

# Load required packages
suppressPackageStartupMessages({
  if (!require("optparse", quietly = TRUE)) {
    stop("Package 'optparse' is required. Install with: install.packages('optparse')")
  }
})

# Color codes for terminal output
color_red <- "\033[0;31m"
color_green <- "\033[0;32m"
color_yellow <- "\033[1;33m"
color_reset <- "\033[0m"

# Logging functions
log_info <- function(msg) {
  cat(sprintf("%s[INFO]%s %s\n", color_green, color_reset, msg))
}

log_error <- function(msg) {
  cat(sprintf("%s[ERROR]%s %s\n", color_red, color_reset, msg), file = stderr())
}

log_warning <- function(msg) {
  cat(sprintf("%s[WARNING]%s %s\n", color_yellow, color_reset, msg))
}

# Function to run system commands with error checking
run_command <- function(cmd, description) {
  log_info(sprintf("Running: %s", description))

  status <- system(cmd, intern = FALSE)

  if (status != 0) {
    log_error(sprintf("Failed: %s", description))
    log_error(sprintf("Command: %s", cmd))
    quit(status = 1)
  }

  log_info(sprintf("Completed: %s", description))
  return(status)
}

# Function to check if a file exists
check_file <- function(filepath, description) {
  if (!file.exists(filepath)) {
    log_error(sprintf("%s not found: %s", description, filepath))
    quit(status = 1)
  }
}

# Function to check if required tools are installed
check_dependencies <- function() {
  tools <- c("fastqc", "STAR", "samtools", "featureCounts")
  missing <- c()

  for (tool in tools) {
    status <- suppressWarnings(system2("which", tool, stdout = FALSE, stderr = FALSE))
    if (status != 0) {
      missing <- c(missing, tool)
    }
  }

  if (length(missing) > 0) {
    log_error(sprintf("Required tools not found: %s", paste(missing, collapse = ", ")))
    log_error("Please install missing dependencies")
    quit(status = 1)
  }
}

# Define command-line options
option_list <- list(
  make_option(c("--read1"), type = "character", default = NULL,
              help = "Path to R1 FASTQ file (gzipped)", metavar = "FILE"),
  make_option(c("--read2"), type = "character", default = NULL,
              help = "Path to R2 FASTQ file (gzipped)", metavar = "FILE"),
  make_option(c("--star-index"), type = "character", default = NULL,
              help = "Path to STAR index directory", metavar = "DIR"),
  make_option(c("--annotation"), type = "character", default = NULL,
              help = "Path to GTF annotation file", metavar = "FILE"),
  make_option(c("--sample-name"), type = "character", default = NULL,
              help = "Sample identifier", metavar = "NAME"),
  make_option(c("--output-dir"), type = "character", default = "results",
              help = "Output directory [default: %default]", metavar = "DIR"),
  make_option(c("--threads"), type = "integer", default = 4,
              help = "Number of threads [default: %default]", metavar = "NUM"),
  make_option(c("--skip-fastqc"), action = "store_true", default = FALSE,
              help = "Skip FastQC step")
)

# Parse command-line arguments
opt_parser <- OptionParser(
  usage = "%prog [options]",
  option_list = option_list,
  description = "\nRNA-seq Analysis Pipeline in R\n\nMinimal pipeline for processing RNA-seq data with FastQC, STAR, and featureCounts.",
  epilogue = paste(
    "Example:",
    "  Rscript rnaseq_pipeline.R \\",
    "    --read1 test_data/reads/sample1_R1.fastq.gz \\",
    "    --read2 test_data/reads/sample1_R2.fastq.gz \\",
    "    --star-index test_data/reference/star_index \\",
    "    --annotation test_data/reference/annotation.gtf \\",
    "    --sample-name sample1 \\",
    "    --output-dir results \\",
    "    --threads 4",
    sep = "\n"
  )
)

opt <- parse_args(opt_parser)

# Validate required arguments
required_args <- c("read1", "read2", "star-index", "annotation", "sample-name")
missing_args <- required_args[sapply(required_args, function(x) is.null(opt[[x]]))]

if (length(missing_args) > 0) {
  log_error(sprintf("Missing required arguments: %s", paste(missing_args, collapse = ", ")))
  print_help(opt_parser)
  quit(status = 1)
}

# Check dependencies
check_dependencies()

# Validate input files
check_file(opt$read1, "Read1 file")
check_file(opt$read2, "Read2 file")
check_file(opt$`star-index`, "STAR index directory")
check_file(opt$annotation, "Annotation file")

# Create output directories
dir.create(opt$`output-dir`, showWarnings = FALSE, recursive = TRUE)
fastqc_dir <- file.path(opt$`output-dir`, "fastqc")
star_dir <- file.path(opt$`output-dir`, "star")
counts_dir <- file.path(opt$`output-dir`, "counts")

if (!opt$`skip-fastqc`) {
  dir.create(fastqc_dir, showWarnings = FALSE, recursive = TRUE)
}
dir.create(star_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(counts_dir, showWarnings = FALSE, recursive = TRUE)

# Print pipeline information
log_info("============================================================")
log_info("RNA-seq Analysis Pipeline Started")
log_info("============================================================")
log_info(sprintf("Sample: %s", opt$`sample-name`))
log_info(sprintf("Read1: %s", opt$read1))
log_info(sprintf("Read2: %s", opt$read2))
log_info(sprintf("STAR index: %s", opt$`star-index`))
log_info(sprintf("Annotation: %s", opt$annotation))
log_info(sprintf("Output: %s", opt$`output-dir`))
log_info(sprintf("Threads: %d", opt$threads))
log_info("============================================================")

# Step 1: FastQC
if (!opt$`skip-fastqc`) {
  log_info("STEP 1: Quality Control with FastQC")

  # Run FastQC on R1
  cmd_r1 <- sprintf("fastqc --outdir %s --threads 2 %s",
                    shQuote(fastqc_dir), shQuote(opt$read1))
  run_command(cmd_r1, "FastQC on R1")

  # Run FastQC on R2
  cmd_r2 <- sprintf("fastqc --outdir %s --threads 2 %s",
                    shQuote(fastqc_dir), shQuote(opt$read2))
  run_command(cmd_r2, "FastQC on R2")

  log_info("FastQC completed")
} else {
  log_info("STEP 1: Skipping FastQC")
}

# Step 2: STAR Alignment
log_info("STEP 2: Alignment with STAR")

cmd_star <- sprintf(
  paste(
    "STAR",
    "--runThreadN %d",
    "--genomeDir %s",
    "--readFilesIn %s %s",
    "--readFilesCommand zcat",
    "--outFileNamePrefix %s",
    "--outSAMtype BAM SortedByCoordinate",
    "--outSAMunmapped Within",
    "--outSAMattributes Standard"
  ),
  opt$threads,
  shQuote(opt$`star-index`),
  shQuote(opt$read1),
  shQuote(opt$read2),
  shQuote(file.path(star_dir, paste0(opt$`sample-name`, ".")))
)

run_command(cmd_star, sprintf("STAR alignment for %s", opt$`sample-name`))

# Index BAM file
bam_file <- file.path(star_dir, paste0(opt$`sample-name`, ".Aligned.sortedByCoord.out.bam"))
cmd_index <- sprintf("samtools index %s", shQuote(bam_file))
run_command(cmd_index, sprintf("Indexing BAM file for %s", opt$`sample-name`))

log_info("STAR alignment completed")

# Step 3: featureCounts
log_info("STEP 3: Gene quantification with featureCounts")

counts_file <- file.path(counts_dir, paste0(opt$`sample-name`, ".counts.txt"))

cmd_counts <- sprintf(
  paste(
    "featureCounts",
    "-T %d",
    "-p",
    "-t exon",
    "-g gene_id",
    "-a %s",
    "-o %s",
    "%s"
  ),
  opt$threads,
  shQuote(opt$annotation),
  shQuote(counts_file),
  shQuote(bam_file)
)

run_command(cmd_counts, sprintf("featureCounts for %s", opt$`sample-name`))

log_info("featureCounts completed")

# Pipeline completion
log_info("============================================================")
log_info("Pipeline completed successfully!")
log_info(sprintf("Results are in: %s", opt$`output-dir`))
log_info(sprintf("Counts file: %s", counts_file))
log_info("============================================================")

# Print output summary
cat("\n")
log_info("Output Files:")
if (!opt$`skip-fastqc`) {
  cat("  FastQC Reports:\n")
  fastqc_files <- list.files(fastqc_dir, pattern = "\\.html$", full.names = TRUE)
  for (f in fastqc_files) {
    cat(sprintf("    %s\n", f))
  }
}
cat("  Alignment:\n")
cat(sprintf("    BAM: %s\n", bam_file))
cat(sprintf("    BAM Index: %s.bai\n", bam_file))
cat(sprintf("    STAR Log: %s\n",
            file.path(star_dir, paste0(opt$`sample-name`, ".Log.final.out"))))
cat("  Counts:\n")
cat(sprintf("    Counts: %s\n", counts_file))
cat(sprintf("    Summary: %s.summary\n", counts_file))
cat("\n")

log_info("Done!")

File: rnaseq_pipeline.py
#!/usr/bin/env python3
"""
RNA-seq Analysis Pipeline
Minimal pipeline: FastQC -> STAR alignment -> featureCounts
"""

import argparse
import subprocess
import sys
from pathlib import Path
import logging


def setup_logging():
    """Configure logging for the pipeline"""
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )
    return logging.getLogger(__name__)


def run_command(cmd, description, logger):
    """Execute a shell command and handle errors"""
    logger.info(f"Running: {description}")
    logger.debug(f"Command: {' '.join(cmd)}")

    try:
        result = subprocess.run(
            cmd,
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        logger.info(f"Completed: {description}")
        return result
    except subprocess.CalledProcessError as e:
        logger.error(f"Failed: {description}")
        logger.error(f"Error: {e.stderr}")
        sys.exit(1)


def fastqc(fastq_file, output_dir, logger):
    """Run FastQC on a FASTQ file"""
    cmd = [
        'fastqc',
        '--outdir', str(output_dir),
        '--threads', '2',
        str(fastq_file)
    ]
    run_command(cmd, f"FastQC on {fastq_file.name}", logger)


def star_align(read1, read2, star_index, sample_name, output_dir, threads, logger):
    """Run STAR alignment"""
    # Create output directory for STAR
    star_output = output_dir / 'star'
    star_output.mkdir(parents=True, exist_ok=True)

    cmd = [
        'STAR',
        '--runThreadN', str(threads),
        '--genomeDir', str(star_index),
        '--readFilesIn', str(read1), str(read2),
        '--readFilesCommand', 'zcat',
        '--outFileNamePrefix', str(star_output / f'{sample_name}.'),
        '--outSAMtype', 'BAM', 'SortedByCoordinate',
        '--outSAMunmapped', 'Within',
        '--outSAMattributes', 'Standard'
    ]
    run_command(cmd, f"STAR alignment for {sample_name}", logger)

    # Index BAM file
    bam_file = star_output / f'{sample_name}.Aligned.sortedByCoord.out.bam'
    cmd_index = ['samtools', 'index', str(bam_file)]
    run_command(cmd_index, f"Indexing BAM file for {sample_name}", logger)

    return bam_file


def feature_counts(bam_file, annotation_gtf, sample_name, output_dir, threads, logger):
    """Run featureCounts for gene quantification"""
    counts_output = output_dir / 'counts'
    counts_output.mkdir(parents=True, exist_ok=True)

    output_file = counts_output / f'{sample_name}.counts.txt'

    cmd = [
        'featureCounts',
        '-T', str(threads),
        '-p',
        '-t', 'exon',
        '-g', 'gene_id',
        '-a', str(annotation_gtf),
        '-o', str(output_file),
        str(bam_file)
    ]
    run_command(cmd, f"featureCounts for {sample_name}", logger)

    return output_file


def main():
    parser = argparse.ArgumentParser(
        description='Minimal RNA-seq analysis pipeline',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Example usage:
  python rnaseq_pipeline.py \\
    --read1 test_data/reads/sample1_R1.fastq.gz \\
    --read2 test_data/reads/sample1_R2.fastq.gz \\
    --star-index test_data/reference/star_index \\
    --annotation test_data/reference/annotation.gtf \\
    --sample-name sample1 \\
    --output-dir results \\
    --threads 4
        """
    )

    parser.add_argument('--read1', type=Path, required=True,
                        help='Path to R1 FASTQ file (gzipped)')
    parser.add_argument('--read2', type=Path, required=True,
                        help='Path to R2 FASTQ file (gzipped)')
    parser.add_argument('--star-index', type=Path, required=True,
                        help='Path to STAR index directory')
    parser.add_argument('--annotation', type=Path, required=True,
                        help='Path to GTF annotation file')
    parser.add_argument('--sample-name', type=str, required=True,
                        help='Sample identifier')
    parser.add_argument('--output-dir', type=Path, default='results',
                        help='Output directory (default: results)')
    parser.add_argument('--threads', type=int, default=4,
                        help='Number of threads (default: 4)')
    parser.add_argument('--skip-fastqc', action='store_true',
                        help='Skip FastQC step')

    args = parser.parse_args()

    # Setup logging
    logger = setup_logging()

    # Validate inputs
    if not args.read1.exists():
        logger.error(f"Read1 file not found: {args.read1}")
        sys.exit(1)
    if not args.read2.exists():
        logger.error(f"Read2 file not found: {args.read2}")
        sys.exit(1)
    if not args.star_index.exists():
        logger.error(f"STAR index directory not found: {args.star_index}")
        sys.exit(1)
    if not args.annotation.exists():
        logger.error(f"Annotation file not found: {args.annotation}")
        sys.exit(1)

    # Create output directory
    args.output_dir.mkdir(parents=True, exist_ok=True)

    logger.info("="*60)
    logger.info("RNA-seq Analysis Pipeline Started")
    logger.info("="*60)
    logger.info(f"Sample: {args.sample_name}")
    logger.info(f"Read1: {args.read1}")
    logger.info(f"Read2: {args.read2}")
    logger.info(f"Output: {args.output_dir}")
    logger.info(f"Threads: {args.threads}")
    logger.info("="*60)

    # Step 1: FastQC
    if not args.skip_fastqc:
        logger.info("STEP 1: Quality Control with FastQC")
        fastqc_dir = args.output_dir / 'fastqc'
        fastqc_dir.mkdir(parents=True, exist_ok=True)
        fastqc(args.read1, fastqc_dir, logger)
        fastqc(args.read2, fastqc_dir, logger)
    else:
        logger.info("STEP 1: Skipping FastQC")

    # Step 2: STAR Alignment
    logger.info("STEP 2: Alignment with STAR")
    bam_file = star_align(
        args.read1,
        args.read2,
        args.star_index,
        args.sample_name,
        args.output_dir,
        args.threads,
        logger
    )

    # Step 3: featureCounts
    logger.info("STEP 3: Gene quantification with featureCounts")
    counts_file = feature_counts(
        bam_file,
        args.annotation,
        args.sample_name,
        args.output_dir,
        args.threads,
        logger
    )

    logger.info("="*60)
    logger.info("Pipeline completed successfully!")
    logger.info(f"Results are in: {args.output_dir}")
    logger.info(f"Counts file: {counts_file}")
    logger.info("="*60)


if __name__ == '__main__':
    main()
